contracts/src/v0.8/automation/upkeeps/LinkAvailableBalanceMonitor.sol

```diff
@@ -135,20 +119,22 @@ contract LinkAvailableBalanceMonitor is ConfirmedOwner, Pausable, AutomationComp
   function sampleUnderfundedAddresses() public view returns (address[] memory) {
     uint16 maxPerform = s_maxPerform;
     uint16 maxCheck = s_maxCheck;
-    uint256 numTargets = s_watchList.length();
-    uint256 idx = uint256(blockhash(block.number - 1)) % numTargets; // start at random index, to distribute load
+    uint256 numTargets = s_watchList.length;
+    uint256 idx = uint256(blockhash(block.number - 1)) % numTargets;
```

Comment:
this is problematic for chainlink automation because the off-chain component expects an upkeep to "be consistent over a small time frame" (there is no rigid definition to adhere to, unfortunately). With the current implementation, the following observations are possible:

| Block | Eligible | performData (upkeep IDs) |
|-------|----------|--------------------------|
| 1     | true     | [1, 50]                  |
| 2     | false    | []                       |
| 3     | true     | [50, 55]                 |
| 4     | false    | []                       |

this will hurt the performance of the system. It would be better if the upkeeps were sampled consistently for a configurable interval. Ex:

```solidity
uint256 idx = uint256(blockhash(block.number - (block.number % interval) - 1)) % numTargets;
```

the complicated part in the middle would yield the following:
| Blockblock.number | block.number - (block.number % interval) - 1 |
|-------------------|----------------------------------------------|
| 100               | 99                                           |
| 101               | 99                                           |
| ...               | ...                                          |
| 109               | 99                                           |
| 110               | 109                                          |
| 111               | 109                                          |

note we have to `- 1` every time to ensure we are never trying to read the blockhash of the current block

`interval` should be a configurable value (different chains with different blocktimes will require different intervals to achieve consistency for the same time window)

finally, we should validate that `interval` >= 255 (because we can't calculate the blockhash of blocks older than 256)

Reply:
Amazing catch, Ryan.